To easily get started with ANNOVAR, there might be some common cases you will consider to do using ANNOVAR. Here, we provide a start to end example for you to follow.
0. Understand the ANNOVAR package and download the dataset/annotation of your interest
When you have requested the ANNOVAR from the website and downloaded it, you will have a file that look like this:
  (base) [wangp5@reslnvhpc0202 ANNOVAR_tutorial]$ ls -lh
  total 156M
  -rw-r--r-- 1 wangp5 wang_lab_hpc 136M Sep 22 21:21 annovar.latest.tar.gz'

1. I have a list of varaints, how do I get population and clinical information for these variants?

2. I want to run ANNOVAR using the new database, or a differnet chromosome, how do I do it?

3. I have a vcf files, how do I run ANNOVAR using my vcf file directly and get the annotation?

4. How do i get the pathogenicity prediction from ANNOVAR, and how do I interpret it?

5. I have a very big vcf file/very large list of variants, how do i run ANNOVAR to process it?
